versions/development/SPEC.md

```diff
@@ -358,6 +359,102 @@ Strings can also contain the following types of escape sequences:
+
+```wdl
+String s = <<<
+  This is a
+  multi-line string!
+>>>
+
+When a multi-line string is evaluated, all whitespace following the opening `<<<`, up to and including a newline (if any), and all whitespace preceeding the closing `>>>`, up to and including a newline (if any), is first stripped.
+
+```wdl# all of these strings evaluate to "hello world"
+String hw1 = <<<hello world>>>
+String hw2 = <<<   hello world   >>>
+String hw3 = <<<   
+    hello world>>>
+String hw4 = <<<   
+    hello world
+    >>>
+```
+
+A multi-line string may contain empty lines and blank lines. An *empty* line contains no characters except a newline. A *blank* line contains only whitespace characers.
+
+After stripping the leading and trailing whitespace, all non-empty lines are used to determine the multi-line string's *common leading whitespace* - the minimum number of whitespace characters occuring before the first non-whitespace character in a line, or the end of the line if the line is blank. This common leading whitespace is stripped from the beginning of all the lines in the multi-line string.
+
+```wdl
+# These strings are all equivalent. The middle lines are empty and so do not count towards the
+# common leading whitespace determination.
+String multi_line_A = <<<
+        this is a
+
+          multi-line string
+>>>
+String multi_line_B = <<<
+  this is a
+
+    multi-line string
+>>>
+String multi_line_C = <<<
+this is a
+
+  multi-line string
+>>>
+
+# This string contains a blank line with 2 spaces, so even though the other two lines are 
+# indented 4 spaces, the common leading whitespace is 2.
+String multi_line_D = <<<
+    this is a
+  
+    multi-line string
+>>>
+```
+
+Each whitespace character is counted once regardless of whether it is a space or tab, so care should be taken when mixing whitespace characters.
+
+A newline character in a multi-line string is left as-is unless it is immediately preceeded by a
+line continuation (`\`), in which case it is removed along with all the whitespace preceeding the next non-whitespace character or end of line (whichever comes first). In other words, a continued line can be considered as part of the previous line, and so it does not contribute to determining the common leading whitespace.
+
+```wdl
+# These two strings are equivalent
+String single_line = "this is a string that contains no newlines"
+String multi_line = <<<
+  this is a \
+    string that \
+ contains no newlines
+>>>
+```
+
+Empty or blank lines may be used in a multi-line string to ensure that it begins/ends with a newline.
+
+```wdl
+# The following strings are equivalent. String `s2` has an empty line, which ensures that it
+# ends in a newline. Even though the first line of `s3` is indented by 6 spaces, the common 
+# leading whitespace is 2 and the string ends in a newline due to the second (blank) line
+# containing 2 spaces.
+String s1 = "    text indented by 4 spaces\n"
+String s2 = <<<
+    text indented by 4 spaces
+
```

Comment:
Since this line is empty, and it doesn't factor into common leading whitespace, wouldn't the 4 spaces on the line above get removed?

Reply:
Oh yea, good point :) So (as the spec is currently written) there is no way to force text to be indented. Are we good with this? Or should we reconsider making empty lines contribute to common leading whitespace determination?

Reply:
@jdidion I don't claim this is great, but at least there's a way

```
        <<<
        \x20 Forced
        \x20 indentation
        >>>

=> "  Forced\n  indentation"
```

Reply:
The more I think about it, the more I dislike the distinction between empty and blank lines. I am going to remove this distinction and see what people think. Can always revert.

Reply:
@jdidion as I understand the revision, a blank line could have some whitespace but less than the "common leading whitespace." Is the procedure supposed to treat the blank line as if it were empty in this case?

Reply:
That's right. If a blank like has `W` whitespace characters, then min(`C`, `W`) whitespace characters are removed, where `C` is common leading whitespace.